pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,568 @@
+	// overloading the cluster. The migration controller handles that. We're merely
+	// optimizing here by not introducing new migration objects we know can't be processed
+	// right now.
+	maxParallelMigrations := int(*c.clusterConfig.GetMigrationConfiguration().ParallelMigrationsPerCluster)
+
+	maxNewMigrations := maxParallelMigrations - data.numActiveMigrations
+	if maxNewMigrations < 0 {
+		maxNewMigrations = 0
+	}
+
+	migrateCount := int(math.Min(float64(maxNewMigrations), float64(len(data.migratableOutdatedVMIs))))
+	migrationCandidates := []*virtv1.VirtualMachineInstance{}
+	if migrateCount > 0 {
+		migrationCandidates = data.migratableOutdatedVMIs[0:migrateCount]
+		log.Log.Infof("workload updated is migrating %d VMIs", migrateCount)
+	}
+
+	evictionCandidates := []*virtv1.VirtualMachineInstance{}
+	if batchDeletionCount > 0 {
+		evictionCandidates = data.evictOutdatedVMIs[0:batchDeletionCount]
+		log.Log.Infof("workload updated is force shutting down %d VMIs", batchDeletionCount)
+	}
+
+	wgLen := len(migrationCandidates) + len(evictionCandidates)
+	wg := &sync.WaitGroup{}
+	wg.Add(wgLen)
+	errChan := make(chan error, wgLen)
+
+	c.migrationExpectations.ExpectCreations(key, migrateCount)
+	for _, vmi := range migrationCandidates {
+		go func(vmi *virtv1.VirtualMachineInstance) {
+			defer wg.Done()
+			createdMigration, err := c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Create(&virtv1.VirtualMachineInstanceMigration{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						virtv1.WorkloadUpdateMigrationAnnotation: "",
+					},
+					GenerateName: "kubevirt-workload-update-",
+				},
+				Spec: virtv1.VirtualMachineInstanceMigrationSpec{
+					VMIName: vmi.Name,
+				},
+			})
+			if err != nil {
+				log.Log.Object(vmi).Reason(err).Errorf("Failed to migrate vmi as part of workload update")
+				c.migrationExpectations.CreationObserved(key)
+				c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedCreateVirtualMachineInstanceMigrationReason, "Error creating a Migration for automated workload update: %v", err)
+				errChan <- err
+				return
+			} else {
+				log.Log.Object(vmi).Infof("Initiated migration of vmi as part of workload update")
+				c.recorder.Eventf(vmi, k8sv1.EventTypeNormal, SuccessfulCreateVirtualMachineInstanceMigrationReason, "Created Migration %s for automated workload update", createdMigration.Name)
+			}
+		}(vmi)
+	}
+
+	for _, vmi := range evictionCandidates {
+		go func(vmi *virtv1.VirtualMachineInstance) {
+			defer wg.Done()
+
+			pod, err := controller.CurrentVMIPod(vmi, c.podInformer)
+			if err != nil {
+
+				log.Log.Object(vmi).Reason(err).Errorf("Failed to detect active pod for vmi during workload update")
+				c.recorder.Eventf(vmi, k8sv1.EventTypeWarning, FailedEvictVirtualMachineInstanceReason, "Error detecting active pod for VMI during workload update: %v", err)
+				errChan <- err
+			}
+
+			err = c.clientset.CoreV1().Pods(vmi.Namespace).Evict(context.Background(),
```

Comment:
Hm, here we may have a problem. KubeVirt is actually blocking the eviction while in the background starting a migration. So while technically you get an error back, practically this succeeded. We simply have no cleaner way right now, since we can't accept that the pod gets a deletion timestamp, but on the other hand can also not return e.g. 201 while still blocking the delete.

Reply:
i can optimize for this. i won't call evict on a VMI that's currently in the process of live migrating. That will at least prevent this controller from spamming out errors as it attempts to retry Evicting

Reply:
fixed